feat(dfu): impl codec

huanmeng-qwq opened the pull request:
for this PR, I've completed the content according to the TODO comments. However, since I directly implemented the codec for Component using Gson as the intermediary, we might not need `TRANSLATION_ARGUMENT`? I'm not sure—maybe this PR is just one possible implementation approach?

kennytv committed:
feat(nbt): update number parsing for 1.21.5

kennytv committed:
chore: move larger chunks of code into separate methods

Gerrygames committed:
1.21.5 component changes

Co-authored-by: MiniDigger | Martin <admin@minidigger.dev>

GliczDev committed:
Add NumberBinaryTag#numberValue

GliczDev committed:
Fix NumberBinaryTag#numberValue javadoc

GliczDev committed:
Add missing `@NotNull` annotation and slightly modify NumberBinaryTag#numberValue javadoc

zml2008 committed:
feat(dfu, nbt-dfu): init

These new modules are intended to allow better integration with the DFU library that drives a growing amount of serialization in modern Minecraft.

zml2008 committed:
wip

zml2008 committed:
release: version 4.20.0

zml2008 committed:
release: prepare for further development on 4.21.0

zml2008 committed:
feat(nbt): Add stream API for compound tags

zml2008 committed:
feat(nbt): Add collectors for list and compound tags

TonytheMacaroni committed:
Use DecorationMap for style builder

tjalp committed:
fix: wrong method name

GliczDev committed:
Update version in `@since` tag

kermandev committed:
feat(nbt): TagStringIO Enhancements

Since adventure release 4.21.0, users can now write any BinaryTag to a String; that left a gap reading them.
There is also another case where it may be useful to capture the rest of the string, for example, in a command argument. (Otherwise users would have to group count)
So this commit:
 * Adds support to read any BinaryTag (disallows trailing like the Compound version)
 * Use a writer for any BinaryTag
 * Adds support for reading embedded any BinaryTag, and CompoundBinaryTag (Allows trailing data, through passed Appendable)
 * Adds support to CharBuffer to take the rest of the buffer.
The special cases exist for compound methods so you can immediately throw on the first failed parsing token, even know BinaryTag's can represent a Compound

zml2008 committed:
Merge pull request #1209 from TonytheMacaroni/style-builder-optimization

Use DecorationMap for style builder

zml2008 committed:
Merge pull request #1188 from GliczDev/numberbinarytag-numbervalue

Add NumberBinaryTag#numberValue

zml2008 committed:
Merge pull request #1208 from KyoriPowered/feat/nbt-streams

feat(nbt): Add stream API for compound/list tags

tjalp committed:
deprecate old method

tjalp committed:
Update Argument.java

Co-authored-by: Riley Park <rileysebastianpark@gmail.com>

kezz committed:
feature(minimessage): Set the target to hold the locale of the translation by default

zml2008 committed:
feat(nbt): Initial implementation of heterogeneous list handling

kezz committed:
Merge pull request #1211 from tjalp/fix/wrong-method-name

fix(minimessage): rename `numeric` to `string`

kezz committed:
feature(api): Skip style builder creation on merge

`mergeStyle` is called a lot, especially in TranslatableComponentRenderer, where there is no merging to be done. By skipping the creation of the builder if the merge is a no-op, we can save some allocations.

kezz committed:
peek to try and determine if we're looking at a key or uuid

this will only break if someone has a minecraft key with a valid uuid value and they don't provide the namespace... but like, meh?

kezz committed:
take into account merges

kezz committed:
make the class final

kezz committed:
Merge pull request #1168 from Gerrygames/update/1.21.5

1.21.5 component changes

kezz committed:
Merge pull request #1216 from KyoriPowered/feature/default-target

feature(minimessage): Set the target to hold the locale of the translation by default

kezz committed:
Merge pull request #1219 from KyoriPowered/fix/dont-create-builder

feature(api): Skip style builder creation on merge

kezz committed:
Merge pull request #1167 from kennytv/feature/snbt-1.21.5

feat(nbt): update number parsing for 1.21.5

zml2008 committed:
Merge pull request #1218 from KyoriPowered/feat/hetero-lists

feat(nbt): Initial implementation of heterogeneous list handling

zml2008 committed:
release: version 4.21.0

zml2008 committed:
release: prepare for further development on 4.22.0

renovate[bot] committed:
chore(deps): Update patch-release dependency changes

renovate[bot] committed:
chore(deps): Update gradle and github actions

TonytheMacaroni committed:
Render translatable arguments and children

TonytheMacaroni committed:
Extract out translatable rendering

TonytheMacaroni committed:
Amend formatting

TonytheMacaroni committed:
Render translatable tags

TonytheMacaroni committed:
Fix rendering hover events with no found translation

zml2008 committed:
Merge pull request #1223 from KyoriPowered/renovate/patch-release-dependency-changes

chore(deps): Update patch-release dependency changes

zml2008 committed:
Merge pull request #1225 from KyoriPowered/renovate/gradle-and-github-actions

chore(deps): Update gradle and github actions

renovate[bot] committed:
chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.38.0

Timongcraft committed:
chore(audience): update javadoc for playSound

add note that stereo sounds will always play as coming from self-emitter

kezz committed:
fix(api): Always fire bossBarNameChanged

Developers use the GlobalTranslator to insert updated content in the resulting component sent to end users, so we shouldn't be checking this component.
The only side effect would be that people who constantly call `name` would have packets repeatedly sent out, which feels like a them problem.

kezz committed:
check if we can translate before doing so

kezz committed:
fix(api): Don't throw an exception if a PointersSupplier has no parent

Stxellxa committed:
Typo fix in PrideTag.java

one of the colours in the Lesbian flag is typod to be D4662A6, and instead should be D462A6

kezz committed:
Merge pull request #1233 from Stxellxa/patch-1

Typo fix in PrideTag.java

kezz committed:
Merge pull request #1230 from KyoriPowered/fix/bb-name-check

fix(api): Always fire bossBarNameChanged

kezz committed:
Merge pull request #1224 from KyoriPowered/renovate/error-prone-monorepo

chore(deps): Update dependency com.google.errorprone:error_prone_core to v2.38.0

huanmeng-qwq committed:
codec

huanmeng-qwq committed:
name

zml2008 commented:
using gson as the intermediary completely defeats the purpose of having dfu codecs, don't do that

huanmeng-qwq committed:
feat: component codec

huanmeng-qwq committed:
update.

huanmeng-qwq commented:
I just submitted the refactored codec implementation for the new Component.

zml2008 commented:
These codecs don't appear to match vanilla output, what is your source for the data structures used here? It also feels like more idiomatic DFU usage would be to use RecordCodecBuilder where possible.

zml2008 commented on dfu/build.gradle.kts:
```diff
@@ -4,6 +4,9 @@ plugins {
 
 dependencies {
   api(libs.dfu8)
+  api(projects.adventureNbtDfu)
```
why?

zml2008 commented on dfu/src/main/java/net/kyori/adventure/dfu/AdventureCodecs.java:
```diff
@@ -1,13 +1,721 @@
+  public static final Codec<Style> STYLE = new StyleCodec();
+  public static final Codec<DataComponentValue> DATA_COMPONENT_VALUE = new DataComponentValueCodec();
+
+  private static <A, S> Codec<S> xmap(Codec<A> codec, Function<? super A, ? extends S> to, Function<? super S, ? extends A> from, String name) {
```
why not use the existing xmap method?

huanmeng-qwq replied:
Because the existing xmap method cannot specify a name, lol.

zml2008 commented:
See build failures too, have you tested this at all? The DFU codecs should have test coverage comparable to other serializer implementations.

huanmeng-qwq committed:
fix styles

huanmeng-qwq committed:
refactor: component codec

huanmeng-qwq committed:
style: fix style

huanmeng-qwq committed:
feat: remove ComponentCodec, refactor DATA_COMPONENT_VALUE

huanmeng-qwq committed:
refactor: click event, hover event, add BINARY_TAG_HOLDER codec, remove DataComponentValueCodec

huanmeng-qwq committed:
refactor: pos, coordinate

huanmeng-qwq committed:
style: fix style

huanmeng-qwq committed:
refactor: component codec

huanmeng-qwq committed:
refactor: index codec

huanmeng-qwq committed:
chore: cleanup

huanmeng-qwq committed:
fix: fix uuid codec

huanmeng-qwq committed:
test: Add tests

huanmeng-qwq commented:
Since the value in `TranslationArgument` is of type Object and could represent multiple numeric types, I can't refactor it to use `RecordCodecBuilder` for implementation.

huanmeng-qwq committed:
style: fix style

huanmeng-qwq committed:
chore: rename test package

huanmeng-qwq commented:
The build failure is caused by `nbt-dfu`.

kezz committed:
use the local parent instance

huanmeng-qwq committed:
feat: partial DATA_COMPONENT_VALUE

kezz committed:
feature(api): Rewrite ComponentFlattener to remove recursion and add a configurable maximum nesting depth

kezz committed:
feature(api): Initial support for native dialog implementations

By providing a method to send dialogs now, we can unify API surface for dialogs before Adventure has a full dialog API.

kermandev committed:
chore(nbt): Change docs to compound

kermandev committed:
chore(nbt): drop extra line

Timongcraft committed:
chore: add version info

kezz committed:
feature(api): Add dialog and custom click events, expand click event with payload system

kezz committed:
Merge pull request #1232 from KyoriPowered/fix/missing-ps

fix(api): Don't throw an exception if a PointersSupplier has no parent

kezz committed:
feature(json): Implement serializing for new click events and int page format

kezz committed:
feature(api): UI sound source and missing source documentation

mworzala committed:
fix: correctly double-box compounds with a single empty key in heterogeneous lists

mworzala committed:
feat: support heterogeneous lists in string nbt

mworzala committed:
chore: pass down heterogeneous flags from TagStringIO

mworzala committed:
fix: formatting

kezz committed:
move docs to Sound and fix checkstyle issues

kezz committed:
address comments, deprecate get method and add reqNonNull methods

kezz committed:
Merge pull request #1238 from KyoriPowered/feature/1.21.6

1.21.6

kezz committed:
Merge pull request #1237 from KyoriPowered/fix/flattening-2

feature(api): Rewrite ComponentFlattener to remove recursion and add a configurable maximum nesting depth

kezz committed:
Merge pull request #1235 from Timongcraft/chore/stereo-sound-nonself-emitter-jd

chore(audience): update javadoc for playSound

kezz committed:
Merge pull request #1239 from kermandev/feat/nbt-stringio

feat(nbt): TagStringIO Enhancements

zml2008 committed:
Merge pull request #1242 from mworzala/feat/hetergeneous_list_improvements

Improved support for hetergeneous lists

zml2008 committed:
Merge branch 'fix-translatable-rendering' into main/4

zml2008 committed:
Merge pull request #1226 from TonytheMacaroni/fix-translatable-rendering

Render translatable arguments and children

zml2008 committed:
release: version 4.22.0

zml2008 committed:
release: prepare for further development on 4.23.0

kezz committed:
fix: Make custom click events hold NBT payloads

kezz committed:
fix: Ensure TagStringIO accepts hetergeneous lists in all reader methods

kezz committed:
fix(key): Ensure keys with invalid namespaces throw the correct exception

Also improve nullability checks on key methods.

zml2008 committed:
Merge pull request #1244 from KyoriPowered/fix/hetero-lists

fix: Ensure TagStringIO accepts hetergeneous lists in all reader methods

zml2008 committed:
Merge pull request #1243 from KyoriPowered/fix/custom-click-event

fix: Make custom click events hold NBT payloads

zml2008 committed:
Merge pull request #1245 from KyoriPowered/fix/string-key

fix(key): Ensure keys with invalid namespaces throw the correct exception

zml2008 committed:
release: version 4.23.0

zml2008 committed:
release: prepare for further development on 4.24.0

kezz committed:
feature(api): Add a method to close an open dialog

Seggan committed:
Add `equals` to `VirtualComponent`

Seggan committed:
Forgot to run checkstyle

Seggan committed:
tfw you forget hashcode

kezz committed:
Merge pull request #1247 from Seggan/main/4

Add `equals` to `VirtualComponent`

kezz committed:
Merge pull request #1246 from KyoriPowered/feature/clear-dialog

feature(api): Add a method to close an open dialog

diogotcorreia committed:
fix: component flattener not popping styles in correct order

Closes #1254

jpenilla committed:
Merge pull request #1255 from diogotcorreia/fix/component-flattener-pop-styles

fix: component flattener not popping styles in correct order

huanmeng-qwq committed:
feat: add DFU codecs for adventure components and styles
